File: shared/src/main/java/com/screentime/shared/firestore/FunctionsErrors.kt
package com.screentime.shared.firestore

import androidx.annotation.StringRes
import com.google.firebase.functions.FirebaseFunctionsException
import com.screentime.shared.R

/**
 * Maps a caught error to a localized [StringRes] instead of showing the
 * server's English-only [Throwable.message] straight in the UI.
 *
 * Prefers the machine-readable `details.reason` tag that functions/src/index.ts
 * attaches to each [HttpsError][com.google.firebase.functions.FirebaseFunctionsException] —
 * the gRPC status code alone is ambiguous (NOT_FOUND alone covers five
 * distinct conditions across the callable functions: invalid code, used
 * code, expired code, missing family, missing pairing).
 */
@StringRes
fun Throwable.toErrorRes(): Int {
    val fe = this as? FirebaseFunctionsException ?: return R.string.error_generic
    val reason = (fe.details as? Map<*, *>)?.get("reason") as? String
    return when (reason) {
        "family_not_found" -> R.string.error_family_not_found
        "not_admin" -> R.string.error_not_admin
        "not_owner" -> R.string.error_not_owner
        "code_invalid" -> R.string.error_code_invalid
        "code_used" -> R.string.error_code_used
        "code_expired" -> R.string.error_code_expired
        "code_invalid_or_expired" -> R.string.error_code_invalid_or_expired
        "tv_already_paired" -> R.string.error_tv_already_paired
        "device_not_paired" -> R.string.error_device_not_paired
        "locked_out" -> R.string.error_locked_out
        else -> when (fe.code) {
            FirebaseFunctionsException.Code.UNAUTHENTICATED -> R.string.error_signin_required
            FirebaseFunctionsException.Code.PERMISSION_DENIED -> R.string.error_permission_denied
            FirebaseFunctionsException.Code.UNAVAILABLE,
            FirebaseFunctionsException.Code.DEADLINE_EXCEEDED -> R.string.error_network
            else -> R.string.error_generic
        }
    }
}

File: web/src/i18n/functionsErrors.ts
import i18n from './i18n';

/**
 * Maps a caught error to a localized message instead of showing the server's
 * English-only Error#message straight in the UI. Mirrors
 * shared/.../FunctionsErrors.kt on the Android side.
 *
 * Prefers the machine-readable `details.reason` tag that functions/src/index.ts
 * attaches to each HttpsError — the gRPC status code alone is ambiguous
 * (NOT_FOUND alone covers five distinct conditions across the callable
 * functions: invalid code, used code, expired code, missing family, missing
 * pairing). The Functions SDK's error class isn't part of the public
 * `firebase/functions` type exports, so this checks structurally instead of
 * with `instanceof`.
 */
interface FunctionsErrorLike {
  code?: string;
  details?: unknown;
}

function isFunctionsErrorLike(error: unknown): error is FunctionsErrorLike {
  return typeof error === 'object' && error !== null && 'code' in error;
}

export function describeFunctionsError(error: unknown): string {
  if (!isFunctionsErrorLike(error)) return i18n.t('errors.generic');
  const reason = (error.details as { reason?: string } | undefined)?.reason;
  switch (reason) {
    case 'family_not_found':
      return i18n.t('errors.familyNotFound');
    case 'not_admin':
      return i18n.t('errors.notAdmin');
    case 'not_owner':
      return i18n.t('errors.notOwner');
    case 'code_invalid':
      return i18n.t('errors.codeInvalid');
    case 'code_used':
      return i18n.t('errors.codeUsed');
    case 'code_expired':
      return i18n.t('errors.codeExpired');
    case 'code_invalid_or_expired':
      return i18n.t('errors.codeInvalidOrExpired');
    case 'tv_already_paired':
      return i18n.t('errors.tvAlreadyPaired');
    case 'device_not_paired':
      return i18n.t('errors.deviceNotPaired');
    case 'locked_out':
      return i18n.t('errors.lockedOut');
    default:
      switch (error.code) {
        case 'functions/unauthenticated':
          return i18n.t('errors.signInRequired');
        case 'functions/permission-denied':
          return i18n.t('errors.permissionDenied');
        case 'functions/unavailable':
        case 'functions/deadline-exceeded':
          return i18n.t('errors.network');
        default:
          return i18n.t('errors.generic');
      }
  }
}
